application/common/components/passwordStore/Multiple.php

```diff
@@ -88,48 +60,70 @@ public function getMeta($employeeId): UserPasswordMeta
+     * with a message detailing which ones succeeded and which ones failed.
      *
      * NOTE: If successful, this will return the UserPasswordMeta returned by
      *       the first password store defined in its list.
      *
      * @param string $employeeId The Employee ID of the user.
      * @param string $password The new password.
      * @return UserPasswordMeta
-     * @throws NotAttemptedException
+     * @throws PasswordStoreException
      * @throws Exception
      * @throws UserNotFoundException
      * @throws AccountLockedException
      * @throws PasswordReuseException
      */
     public function set($employeeId, $password): UserPasswordMeta
     {
-        $this->assertAllBackendsAreAvailable($employeeId, 'set the password');
-        $numSuccessfullySet = 0;
         $responses = [];
+        $successes = [];
+        $errors = [];
         foreach ($this->passwordStores as $passwordStore) {
             try {
                 $responses[] = $passwordStore->set($employeeId, $password);
-                $numSuccessfullySet++;
+                $successes[] = $passwordStore->getDisplayName();
             } catch (Exception $e) {
                 if ($e instanceof PasswordReuseException) {
                     // Be aware that this does not include information about how many backends the password
                     // was successfully changed in
                     throw $e;
                 }
-                throw new Exception(sprintf(
-                    'Failed to set the password using %s after successfully '
-                    . 'setting it in %s other password store(s). Error: %s',
-                    \get_class($passwordStore),
-                    $numSuccessfullySet,
-                    $e->getMessage()
-                ), 1498162884, $e);
+
+                \Yii::error([
+                    'action' => 'set password',
+                    'status' => 'error',
+                    'passwordStore' => $passwordStore->getDisplayName(),
+                    'message' => $e->getMessage(),
+                ]);
+                $errors[] = $passwordStore->getDisplayName();
+            }
+        }
+
+        if (count($errors) > 0) {
+            if (count($successes) > 0) {
+                $errorMessage = \Yii::t(
+                    'app',
+                    'Multiple.SetPartialSuccess {successes} {errors}',
+                    [
+                        'successes' => implode(', ', $successes),
+                        'errors' => implode(', ', $errors),
+                    ]
+                );
+            } else {
+                $errorMessage = \Yii::t(
+                    'app',
+                    'Multiple.SetFailed {errors}',
+                    [
+                        'errors' => implode(', ', $errors),
+                    ]
+                );
             }
+            throw new PasswordStoreException($errorMessage, 1498162884);
         }
+
         return $responses[0];
```

Comment:
This isn't a change made by this PR, but is there anything checking that there is at least one password store?

Reply:
Yes, in `init()`
```php
        if (empty($this->passwordStoresConfig)) {
            throw new InvalidArgumentException(
                'You must provide config for at least one password store.',
                1498162679
            );
        }
```

Reply:
Okay, great. Thanks.